Add a build array method that will throw if allocations fails

bmcdonald3 opened the pull request:
In Arkouda, out of memory issues have been a problem when allocating
large arrays, so this PR adds support for throwing if an array
allocation fails, as opposed to the previous behavior of halting,
which would crash the server in Arkouda's case.

This is supported on both block and default rectangular arrays, as
those are all that is used in Arkouda.

This PR will only work in fixed-heap configurations where overcommit
is disabled.

benharsh commented on modules/dists/BlockDist.chpl:
```diff
@@ -952,6 +967,54 @@ proc BlockDom.dsiBuildArray(type eltType, param initElts:bool) {
+      var callPostAlloc = false;
+      var data = _ddata_allocate_noinit_nocheck(eltType, locSize, callPostAlloc);
+      if data == nil then
+        throw new Error("Could not allocate memory");
```
Might be worth adding a TODO somewhere that this ``Error`` should eventually be a different type that users could more easily distinguish.

benharsh commented on modules/internal/ChapelDomain.chpl:
```diff
@@ -1493,6 +1493,41 @@ module ChapelDomain {
+      and block domains.
+    */
+    pragma "no copy return"
+    proc tryBuildArray(type eltType, param initElts=true) throws {
```
Should this start off with ``@unstable`` for now?

benharsh commented on test/memory/benmcd/oomCheck.chpl:
```diff
@@ -0,0 +1,7 @@
+
+const dom = {0..#size};
+
+var myArr = dom.tryBuildArray(int);
```
I think there should be a test that catches the error and responds to it in some way.

I think it would also be good if there was a test that looked for memory leaks in a situation like this, to see if we leak any of the Block-related classes somehow.

bmcdonald3 replied:
Looks like there are memory leaks! Will work to get those fixed, good catch.

benharsh commented on modules/internal/ChapelDomain.chpl:
```diff
@@ -1493,6 +1493,41 @@ module ChapelDomain {
+      error when attempting to access the array.
+
+      This method is currently supported on both default rectangular
+      and block domains.
```
Could you add a ``compilerError`` with a corresponding test that triggers the error with an incompatible domain type?

benharsh replied:
Another way to handle this would be ``canResolveMethod``, but this is fine too.

bmcdonald3 commented:
Alright, I believe that I have implemented all of the feedback and added the testing requested. It turned out there wasn't any memory actually leaking, but when the allocation failed, I still was tracking that allocation in the runtime, so I had to make sure that the allocation succeeded before tracking the memory.

benharsh commented on modules/internal/ChapelDomain.chpl:
```diff
@@ -1512,12 +1512,16 @@ module ChapelDomain {
+    @unstable("tryCreateArray() is subject to change in the future.")
+    proc tryCreateArray(type eltType) throws {
+      if !this.isDefaultRectangular() && !this.isBlock() then
+        compilerError("'tryBuildArray' is only supported on " +
```
This still says 'tryBuildArray', rather than 'tryCreateArray'.

benharsh replied:
Also, might as well change the "doi" method names.

bmcdonald3 replied:
I wasn't sure about the `doi` method names, since those are essentially copy-pastes of the `dsiBuildArray` method, so keeping the `build` naming there seemed more in line with that. With that in mind, do you still prefer switching that to `doiCreateArray`?

benharsh replied:
I think the behavior is sufficiently different to warrant the name change to ``doiTryCreateArray``. That way it's the same name at every level and there's less potential confusion when discussing or searching for the relevant methods.

benharsh reviewed (approved):
Just a couple of things, otherwise, looks good! No need for a re-review once you fix them.

ronawho commented on modules/internal/ChapelDomain.chpl:
```diff
@@ -1493,6 +1493,46 @@ module ChapelDomain {
+    pragma "no copy return"
+    @unstable("tryCreateArray() is subject to change in the future.")
+    proc tryCreateArray(type eltType) throws {
+      if !this.isDefaultRectangular() && !this.isBlock() then
```
Instead of adding `isBlock` I think you can do `isSubtype(this.dist.type, Block);` (or something like that, we do a similar check in Arkouda's `contiguousIndices`)

bradcray commented:
Nice!  I'm behind on your issues so am asking the lazy question here:  What's the level of effort required to get this being used by Arkouda itself?

bmcdonald3 commented:
The level of effort for adding this into Arkouda symbol table entry creations (~90% of Arkouda large array allocations) is very low and really just requires an update to the `createSymEntry` array creation wrapper and a few other places to handle array initialization (since we can't throw in initializers).

For every single array allocation, that is a bit higher effort, where I am still hitting some issues that I don't understand entirely. As far as things like `scan`s go, I had patch to enable throwing scans, but there isn't the fancy compiler magic to make the syntax nice, so that is relatively ugly.

I think the plan that Elliot and I have discussed is that we are going to get just the array creation code wired through Arkouda and then see if that helps resolve the issue and reassess from there, since there are quite a few open-ended questions like I've mentioned regarding wider support and the symbol table entries are the vast majority of large array allocations.
